### PR TITLE
Fix secret/configmap management for terminated pods

### PR DESCRIPTION
**Pull request by oxddr**
Ref #74412 

This PR is based on work of @wojtek-t in #74730 

```release-note
Kubelet no longer watches configmaps and secrets for terminated pods, in worst scenario causing it to not be able to send other requests to kube-apiserver
```

### CODE REVIEW

File: pkg/kubelet/util/manager/cache_based_manager_test.go [hunk 1]
```diff
@@ -429,6 +429,37 @@ func TestCacheInvalidation(t *testing.T) {
...
+			return 0
+		}
+		return item.refCount
+	}
+
+	manager.RegisterPod(podWithSecrets("ns1", "name1", s1))
+	assert.Equal(t, 1, refs("ns1", "s1"))
+	manager.RegisterPod(podWithSecrets("ns1", "name1", s1))
+	assert.Equal(t, 1, refs("ns1", "s1"))
+
```
Comment (liggitt) [inline pkg/kubelet/util/manager/cache_based_manager_test.go]: suggest registering a second pod and making sure the refcount is 2

File: pkg/kubelet/util/manager/cache_based_manager_test.go [hunk 2]
```diff
@@ -429,6 +429,37 @@ func TestCacheInvalidation(t *testing.T) {
...
+	manager.UnregisterPod(podWithSecrets("ns1", "name1", s1))
+	assert.Equal(t, 0, refs("ns1", "s1"))
+	manager.UnregisterPod(podWithSecrets("ns1", "name1", s1))
+	assert.Equal(t, 0, refs("ns1", "s1"))
+}
```
Comment (liggitt) [inline pkg/kubelet/util/manager/cache_based_manager_test.go]: and then verifying the ref count is 1 after unregistering the single pod multiple times

**Comment by liggitt** [comment]
does a pod with a single crashlooping container ever enter failed or succeeded state?

we use secretmanager to get imagepullsecrets which could be needed for pulling the container image on a restart:

https://github.com/kubernetes/kubernetes/blob/dcbe588f76127cd7a3360e70f2260cd7c5ebbc73/pkg/kubelet/kubelet_pods.go#L858-L866

**Comment by liggitt** [comment]
answering my own question, a restarting container keeps a pod in Running:

https://github.com/kubernetes/kubernetes/blob/dcbe588f76127cd7a3360e70f2260cd7c5ebbc73/pkg/apis/core/types.go#L2134-L2136

and succeeded/failed is terminal:

https://github.com/kubernetes/kubernetes/blob/dcbe588f76127cd7a3360e70f2260cd7c5ebbc73/pkg/kubelet/kubelet_pods.go#L1364-L1372

**Comment by oxddr** [comment]
I think I need an approval from you, @yujuhong.

**Comment by yujuhong** [comment]
Please add a release note.

**Comment by wojtek-t** [comment]
/ok-to-test

@yujuhong - I copied the release-note from my PR.

**Pull request merged**